Make SharedMemoryHandle non-copyable

donny-dont opened the pull request:
#### 00afd6a353256ecc8010e64ac280ca4e1159825b
<pre>
Make SharedMemoryHandle non-copyable
<a href="https://bugs.webkit.org/show_bug.cgi?id=256539">https://bugs.webkit.org/show_bug.cgi?id=256539</a>

Reviewed by Kimmo Kinnunen.

Add `WTF_MAKE_NONCOPYABLE` to `SharedMemoryHandle`. Enumerate
constructors for the class.

Move common implementation of `SharedMemoryHandle` into common .cpp.

* Source/WebKit/Platform/SharedMemory.cpp:
(WebKit::SharedMemoryHandle::SharedMemoryHandle):
(WebKit::SharedMemoryHandle::isNull const):
* Source/WebKit/Platform/SharedMemory.h:
* Source/WebKit/Platform/cocoa/SharedMemoryCocoa.cpp:
(WebKit::SharedMemoryHandle::isNull const): Deleted.
* Source/WebKit/Platform/unix/SharedMemoryUnix.cpp:
(WebKit::SharedMemoryHandle::isNull const): Deleted.
* Source/WebKit/Platform/win/SharedMemoryWin.cpp:
(WebKit::SharedMemoryHandle::isNull const): Deleted.
* Source/WebKit/Shared/IPCStreamTester.cpp:
(WebKit::IPCStreamTester::syncMessageReturningSharedMemory1):
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::JSSharedMemory::createHandle):

Canonical link: <a href="https://commits.webkit.org/263897@main">https://commits.webkit.org/263897@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cbd8ee440a0beb863a9a14cd1c43af30ae3df8e8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6029 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6207 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6393 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7585 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6382 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6425 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6163 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9269 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6138 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6166 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5459 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7645 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3642 "Passed tests") | | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13350 "1 flakes 1 missing results 188 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5508 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5516 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/7735 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5973 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4894 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5399 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1434 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9533 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5765 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebKit/Platform/SharedMemory.h:
```diff
@@ -70,6 +71,12 @@ class SharedMemoryHandle {
 
+    static SharedMemoryHandle create(Type&&, size_t);
+
+    SharedMemoryHandle() = default;
```
I think later on the default constructor, clear() and isNull() should be removed.
That might be appropriate for this commit, or then some other commit..

donny-dont replied:
Ok so I removed clear in https://github.com/WebKit/WebKit/commit/62bea872d1f186dcb512d2296f2e5cce1f499f94 but removing the default constructor which would be required for removing `isNull` is non-trivial and would touch a lot of the classes that needed to be modified to get all the move semantics changes.

kkinnunen-apple commented on Source/WebKit/Platform/SharedMemory.h:
```diff
@@ -70,6 +71,12 @@ class SharedMemoryHandle {
         Win32Handle;
 #endif
 
+    static SharedMemoryHandle create(Type&&, size_t);
```
There's no benefit of having create function that returns the object. That is just a normal constructor.

donny-dont replied:
Ok you had mentioned having a `create` method for `SharedMemoryHandle` but I can drop this.

kkinnunen-apple commented on Source/WebKit/Platform/SharedMemory.h:
```diff
@@ -85,6 +92,8 @@ class SharedMemoryHandle {
 
 private:
+    SharedMemoryHandle(Type&&, size_t);
+
```
I think m_handle should be able to be non-mutable as of now..

donny-dont replied:
I wasn't able to remove `mutable` because of the generated messages.

donny-dont committed:
Make SharedMemoryHandle non-copyable
https://bugs.webkit.org/show_bug.cgi?id=256539

Reviewed by Kimmo Kinnunen.

Add `WTF_MAKE_NONCOPYABLE` to `SharedMemoryHandle`. Enumerate
constructors for the class.

Move common implementation of `SharedMemoryHandle` into common .cpp.

* Source/WebKit/Platform/SharedMemory.cpp:
(WebKit::SharedMemoryHandle::SharedMemoryHandle):
(WebKit::SharedMemoryHandle::isNull const):
* Source/WebKit/Platform/SharedMemory.h:
* Source/WebKit/Platform/cocoa/SharedMemoryCocoa.cpp:
(WebKit::SharedMemoryHandle::isNull const): Deleted.
* Source/WebKit/Platform/unix/SharedMemoryUnix.cpp:
(WebKit::SharedMemoryHandle::isNull const): Deleted.
* Source/WebKit/Platform/win/SharedMemoryWin.cpp:
(WebKit::SharedMemoryHandle::isNull const): Deleted.
* Source/WebKit/Shared/IPCStreamTester.cpp:
(WebKit::IPCStreamTester::syncMessageReturningSharedMemory1):
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::JSSharedMemory::createHandle):

Canonical link: https://commits.webkit.org/263897@main
